ambry-clustermap/src/main/java/com.github.ambry.clustermap/AmbryReplicaSyncUpManager.java

```diff
@@ -105,18 +108,62 @@ public boolean isSyncUpComplete(ReplicaId replicaId) {
    * unblock external service waiting on this latch.
    */
   @Override
-  public void onBootstrapComplete(String partitionName) {
-    partitionToBootstrapSuccess.put(partitionName, true);
-    countDownLatch(partitionName);
+  public void onBootstrapComplete(ReplicaId replicaId) {
+    partitionToBootstrapSuccess.put(replicaId.getPartitionId().toPathString(), true);
+    replicaToLagInfos.remove(replicaId);
+    countDownLatch(partitionToBootstrapLatch, replicaId.getPartitionId().toPathString());
   }
 
   /**
    * {@inheritDoc}
    * This method will count down latch and terminates bootstrap.
    */
   @Override
-  public void onBootstrapError(String partitionName) {
-    countDownLatch(partitionName);
+  public void onBootstrapError(ReplicaId replicaId) {
+    replicaToLagInfos.remove(replicaId);
+    countDownLatch(partitionToBootstrapLatch, replicaId.getPartitionId().toPathString());
+  }
+
+  @Override
+  public void initiateDisconnection(ReplicaId replicaId) {
+    partitionToDisconnectionLatch.put(replicaId.getPartitionId().toPathString(), new CountDownLatch(1));
+    partitionToDisconnectionSuccess.put(replicaId.getPartitionId().toPathString(), false);
+    // once disconnection is initiated, local replica won't receive any PUT/DELETE/TTLUpdate. All remote replicas should
+    // be able to eventually catch with local replica. Hence, we set acceptable lag threshold to 0.
+    replicaToLagInfos.put(replicaId, new LocalReplicaLagInfos(replicaId, 0, ReplicaState.OFFLINE));
+  }
+
+  @Override
+  public void waitDisconnectionCompleted(String partitionName) throws InterruptedException {
+    CountDownLatch latch = partitionToDisconnectionLatch.get(partitionName);
+    if (latch == null) {
+      logger.error("Partition {} is not found for disconnection", partitionName);
+      throw new StateTransitionException("No disconnection latch is found for partition " + partitionName,
+          StateTransitionException.TransitionErrorCode.ReplicaNotFound);
```

Comment:
this makes waitDisconnectionCompleted method non-reentrant? Is this intentional?

Reply:
will discuss this offline (my first thought is this method is not supposed to be reentrant)